### ISSUE TITLE
Rendering Voxels causes 'floating point exception' crash in release build

### ISSUE THREAD

**Issue by JoeAnonymous** [fabiangreffrath/woof#1816]
With a wad file containing voxels the game will crash occasionally when rendering them. 
The text in the terminal states: 'Floating point exception (core dumped)'
Compiling the debug build stops the crashing all together

**Comment by rfomin** [comment]
That's strange, voxel code dosen't use float point at all.

What OS are you using? Also, what is this wad file?

**Comment by fabiangreffrath** [comment]
> This is strange, voxel code does not use floating point at all.

https://stackoverflow.com/questions/16928942/why-does-integer-division-by-zero-result-in-a-floating-point-exception

**Comment by JoeAnonymous** [comment]
> This is strange, voxel code does not use floating point at all.
> 
> What OS are you using? Also, what is this wad file?

I'm running Xubuntu LTS 22.04

The wad file was 'made' by me, but the assets were from voxel doom 2, which is a .pk3 file since it was originally designed for GZDoom

**Comment by rfomin** [comment]
> The wad file was 'made' by me, but the assets were from voxel doom 2, which is a .pk3 file since it was originally designed for GZDoom

You can just run this pk3 directly, it's supported by Woof.
```
woof -file VoxelDoom_v_2.3.pk3
```

I ran some long demos on Linux with voxels, but I still can't catch the crash. Can you reproduce the problem with this AppImage: https://github.com/fabiangreffrath/woof/actions/runs/10218188573/artifacts/1769861565

**Comment by JoeAnonymous** [comment]
> I ran some long demos on Linux with voxels, but I still can't catch the crash. Can you reproduce the problem with this AppImage: https://github.com/fabiangreffrath/woof/actions/runs/10218188573/artifacts/1769861565

I managed to make it crash with both the .wad and .pk3 file. The command I used was this:
Woof-14.5.0-Linux.appimage -iwad DOOM2.WAD -file pwad/<the file containing the voxels>

When running the game, all I had to do was let a demo run for the crash to happen. I also ran the appimage without the iwad flag and it didn't change the results

I store the IWAD's in /usr/local/share/games/doom, and the PWAD's are in a subfolder

**Comment by rfomin** [comment]
I've made changes here: https://github.com/fabiangreffrath/woof/pull/1818

Please test this AppImage: https://github.com/fabiangreffrath/woof/actions/runs/10233383473/artifacts/1773071338

**Comment by JoeAnonymous** [comment]
> I've made changes here: #1818
> 
> Please test this AppImage: https://github.com/fabiangreffrath/woof/actions/runs/10233383473/artifacts/1773071338

Now it works without crashes, thank you very much

**Comment by fabiangreffrath** [comment]
If this crash is really caused by a division by zero, I don't see how this could be mitigated by switching from an ASM implementation back to the `/` operator.

**Comment by fabiangreffrath** [comment]
And I don't understand why this case isn't caught by the `if ((abs(a) >> 14) >= abs(b))` check.

**Comment by rfomin** [comment]
> If this crash is really caused by a division by zero, I don't see how this could be mitigated by switching from an ASM implementation back to the `/` operator.

I think it's caused by the "quotient is too large for the designated register" exception, which is the same as "source operand (divisor) is 0". See https://www.felixcloutier.com/x86/idiv#64-bit-mode-exceptions

**Comment by fabiangreffrath** [comment]
I see. Is there some way we can catch this? I wonder how built-in operators can handle that.

**Comment by rfomin** [comment]
> Is there some way we can catch this?

Yes, but since I can't reproduce it, I'm not sure.

> I wonder how built-in operators can handle that.

In this case, the built-in operators are 64-bit.

**Comment by fabiangreffrath** [comment]
> In this case, the built-in operators are 64-bit.

That'd be `idivq` then? Cold we use this and cast the result back to `int32_t`? I'm not sure...

**Comment by rfomin** [comment]
> > In this case, the built-in operators are 64-bit.
> 
> That'd be `idivq` then? Cold we use this and cast the result back to `int32_t`? I'm not sure...

No, it's all about dividing a 64-bit number by a 32-bit divisor, because this operation is much faster than 64-bit division on many CPUs that aren't that old.

The Linux kernel contains some relevant code: https://github.com/torvalds/linux/blob/master/arch/x86/include/asm/div64.h

**Comment by gendlin** [comment]
Not sure if related to the bug in question, but I think
```c
int32_t lo = a;
int32_t hi = a >> 32;
```
would be better changed to
```c
uint32_t lo = a;
uint32_t hi = (uint64_t)a >> 32;
```
This has less reliance on implementation-defined behavior [[ref](https://rgambord.github.io/c99-doc/sections/6/3/1/3/index.html)] and will generate a logical shift (`shr`) rather than an arithmetic shift (`sar`), which matches the generated assembly for `_div64` [[ref](https://godbolt.org/z/Ebz1q3xsr)].

However, I don't see how this explains the crash unless there is a miscompilation going on (no crash in debug would support that).

**Comment by rfomin** [comment]
> or use a union like the referenced linux code (probably the best way)

Yes, your version looks better.

> However, I don't see how this explains the crash unless there is a miscompilation going on (no crash in debug would support that).

I don't know what's wrong, even the previous version works for me with Clang/GCC on Win and Linux.

**Comment by gendlin** [comment]
If @JoeAnonymous is still around I'd be curious if he can reproduce the crash when uncapped framerate is turned off. Also, what resolution and framerate limit if any was being used, and actual framerate using SHOWFPS cheat as those will all affect the voxel calcs AFAIK and may help us reproduce.

**Comment by JoeAnonymous** [comment]
> If @JoeAnonymous is still around I'd be curious if he can reproduce the crash when uncapped framerate is turned off. Also, what resolution and framerate limit if any was being used, and actual framerate using SHOWFPS cheat as those will all affect the voxel calcs AFAIK and may help us reproduce.

it was already off, actually.
as for resolution, it was 200%.
but the other values i don't know, i will have to check those

**Comment by gendlin** [comment]
Thanks. There's a candidate fix in #1832 that preserves voxel optimizations which you may also want to try.
